rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -119,44 +135,151 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
+      // Find indices in z-direction for each of the vertices. If link is not 
+      // within the range of one of the columns, set to lowest or highest two.
+      int idx_k[8] = {0, static_cast<int>(vertical_spacing_factors_.size()) - 1, 0,
+                      static_cast<int>(vertical_spacing_factors_.size()) - 1, 0,
+                      static_cast<int>(vertical_spacing_factors_.size()) - 1, 0,
+                      static_cast<int>(vertical_spacing_factors_.size()) - 1};
+
+      for (int i = 0; i < 4; i++) {
+        if (vertical_factors_columns[i] < 0) {
+          // Link z-position below lowest grid point of that column.
+          idx_k[2 * i + 1] = 1;
+        } else if (vertical_factors_columns[i] >= 1) {
+          // Link z-position above highest grid point of that column.
+          idx_k[2 * i] = vertical_spacing_factors_.size() - 2;
+        } else {
+          // Link z-position between two grid points in that column.
+          for (int j = 0; j < vertical_spacing_factors_.size() - 1; j++) {
+            if (vertical_spacing_factors_[j] <= vertical_factors_columns[i] && 
+                vertical_spacing_factors_[j + 1] > vertical_factors_columns[i]) {
+              idx_k[2 * i] = j;
+              idx_k[2 * i + 1] = j + 1;
+              break;
+            }
+          }
+        }
+      }
+
+      // Extract the wind velocities corresponding to each vertex.
+      math::Vector3 wind_at_vertices[8];
+      for (int i = 0; i < 8; i++) {
+        wind_at_vertices[i].x = u_[idx_i[i] + idx_j[i] * n_x_ + idx_k[i] * n_x_
+                                                                        * n_y_];
+        wind_at_vertices[i].y = v_[idx_i[i] + idx_j[i] * n_x_ + idx_k[i] * n_x_
+                                                                        * n_y_];
+        wind_at_vertices[i].z = w_[idx_i[i] + idx_j[i] * n_x_ + idx_k[i] * n_x_
+                                                                        * n_y_];
+      }
+
+      // Extract the relevant coordinate of every point needed for trilinear 
+      // interpolation (first z-direction, then x-direction, then y-direction).
+      float interpolation_points[14];
+      for (int i = 0; i < 14; i++) {
+        if (i < 8) {
+          interpolation_points[i] = (
+            top_z_[idx_i[i] + idx_j[i] * n_x_] - bottom_z_[idx_i[i] + idx_j[i] * n_x_])
+            * vertical_spacing_factors_[idx_k[i]] + bottom_z_[idx_i[i] + idx_j[i] * n_x_];
+        } else if (i >= 8 && i < 12) {
+          interpolation_points[i] = min_x_ + res_x_ * idx_i[2 * (i - 8)];
+        } else {
+          interpolation_points[i] = min_y_ + res_y_ * idx_j[i - 9];
+        }
+      }
+
+      wind_velocity = TrilinearInterpolation(
+        link_position, wind_at_vertices, interpolation_points);
+    } else {
+      // Set the wind velocity to the default constant value specified by user.
+      double wind_speed = wind_speed_mean_;
```

Comment:
same here, not sure if you need to introduce the additional `wind_speed` variable.